Fivetran tutorial (#1937)
astronomer/docs: learn/airflow-fivetran.md

2. Commit at least one change to the repository. The content of your commit does not matter for this tutorial.

## Step 3: Connect Fivetran to a destination

Fivetran needs at least one destination to be configured in order to create syncs. A destination is a relational database where your ingested data will be loaded into.

1. Log in to your [Fivetran account](https://www.fivetran.com/). 

2. Click **Destinations** in the left menu, then click **ADD DESTINATION** in the upper right corner.

3. Choose either **Connect your destination** or **I don't have one**. For this tutorial you can either connect to an existing data warehouse by following the [relevant Fivetran documentation](https://fivetran.com/docs/databases), or use a [Fivetran-managed BigQuery service](https://fivetran.com/docs/destinations/bigquery/managed-bigquery). In this tutorial we will use Fivetran's managed BigQuery service by selecting **I don't have one** and then clicking **CONTINUE SETUP**.

4. Configure your destination connection. You can choose any configuration. Astronomer recommends using UTC as your timezone in all data tools as a best practice. Click **SAVE & TEST** to save your destination.

> 5. Click **Continue** to get back to the list of your destinations.

Reviewer: ```suggestion
5. Click **Continue** to return to the list of your destinations.
```

## Step 4: Configure a Fivetran connector

> Fivetran needs at least one [connector](https://fivetran.com/docs/getting-started/fivetran-dashboard/connectors) to be configured in order to create syncs. A Fivetran connector reaches out to a specific data source, receives data from it and writes it to your destination. View the Fivetran website for an [up-to-date list of connectors](https://www.fivetran.com/connectors).

Reviewer: ```suggestion
Fivetran needs at least one configured [connector](https://fivetran.com/docs/getting-started/fivetran-dashboard/connectors) to create syncs. A Fivetran connector reaches out to a specific data source, receives data from it, and writes it to your destination.
```

1. In Fivetran, click **Connectors** in the left menu and then click **ADD CONNECTOR** in the upper right corner.

2. Select the GitHub Connector and click **CONTINUE SETUP**.

3. Configure the GitHub Connector: 
    

>     **Destination schema**: `in_github`
>     **Authentication mode**: Either [OAuth or a Personal Access Token](https://fivetran.com/docs/applications/github/setup-guide), then click **AUTHORIZE**. In this tutorial, we use OAuth authentication. 

Reviewer: ```suggestion
    - **Destination schema**: `in_github`
    - **Authentication mode**: Either [OAuth or a Personal Access Token](https://fivetran.com/docs/applications/github/setup-guide), then click **AUTHORIZE**. In this tutorial, we use OAuth authentication. 
```

4. Authenticate Fivetran to your Github Account.

5. In **Sync Mode** select **Sync Specific Repositories**, then in **Repositories** select the `airflow-fivetran-tutorial` repository.

> 6. Click **SAVE & TEST**. After your connection has been tested click **CONTINUE**.

Reviewer: ```suggestion
6. Click **SAVE & TEST**. After the test succeeds, click **CONTINUE**.
```

> 7. Click **Start Sync** to start your initial sync. This initial synchronization will load all historic metadata from your GitHub repository and has to be completed in Fivetran for the sync to become active. Once the Fivetran sync is active, you can set the sync frequency under the **Setup** tab or run the sync using the Fivetran API in Airflow.

Reviewer: ```suggestion
7. Click **Start Sync** to start the required initial sync. This initial synchronization will load all historic metadata from your GitHub repository. Once the Fivetran sync is active, you can set the sync frequency under the **Setup** tab or run the sync using the Fivetran API in Airflow.
```

## Step 5: Generate a Fivetran API key

> You have now created a Fivetran [sync](https://fivetran.com/docs/getting-started/syncoverview) that will extract new metadata from a GitHub repository and load it into a relational database on a time-based schedule. But to run the sync every time a specific event happens in your data ecosystem, you need to use Airflow for orchestration. 

Reviewer: ```suggestion
You now have a Fivetran [sync](https://fivetran.com/docs/getting-started/syncoverview) that will extract new metadata from a GitHub repository and load it into a relational database on a time-based schedule. But to run the sync every time a specific event happens in your data ecosystem, you need to use Airflow for orchestration. 
```

4. Click **Test** to test your connection. Once you see `Connection tested successfully`, click **Save**.

## Step 7: Create your Airflow DAG

For this tutorial you will create a DAG that triggers your Fivetran sync to ingest the GitHub repo metadata to your destination.

1. Open your `astro-fivetran-project` in a code-editor.

2. In your `dags` folder add a new Python file called `my_fivetran_dag.py`. 

3. Copy and paste the following DAG code into the file:

    <CodeBlock language="python">{fivetran_tutorial_dag_1}</CodeBlock>

    This DAG contains three tasks: 

    - The `upstream` task runs before the Fivetran sync job is run. This task could contain a sensor or deferrable operator waiting for an action to be completed in another data tool.
    - The `run_fivetran_sync` task uses the FivetranOperatorAsync to trigger the Fivetran connector specified as `FIVETRAN_CONNECTOR_ID` as soon as the `upstream` task has completed successfully.

>     - The `downstream` task runs after the Fivetran sync job has finished. Commonly, tasks containing data transformations on the data loaded into your Fivetran destination will be set as downstream tasks.

Reviewer: ```suggestion
    - The `downstream` task runs after the Fivetran sync job has finished. This task could contain data transformations on the data loaded into your Fivetran destination.
```

4. Update the `FIVETRAN_CONNECTOR_ID` variable with the ID of your connector. You can find the ID of your connector in the Fivetran UI under the **Setup** tab:

    ![Fivetran connector ID](/img/guides/fivetran_connector_id.png)

5. Add your GitHub username to the `GITHUB_REPOSITORY` variable.

6. Save your DAG file with the changed variable names.

> The FivetranOperatorAsync is one of many [deferrable operators](deferrable-operators.md). Instead of taking up a worker slot, these operators will hand their task to the Airflow Triggerer component while waiting for a condition to be fulfilled. For longer running tasks, this can result in cost savings and greater scalability as more worker slots area available.

Reviewer: ```suggestion
The FivetranOperatorAsync is one of many [deferrable operators](deferrable-operators.md). Instead of taking up a worker slot, these operators will hand their task to the Airflow Triggerer component while waiting for a condition to be fulfilled. For longer running tasks, this can result in cost savings and greater scalability as more worker slots are available.
```

2. Commit at least one change to the repository. The content of your commit does not matter for this tutorial.

## Step 3: Connect Fivetran to a destination

Fivetran needs at least one destination to be configured in order to create sync jobs. A destination is a relational database where your ingested data will be loaded into.

1. Log into your [Fivetran account](https://www.fivetran.com/). 

2. Click on **Destinations** in the sidebar on the left side of the screen and then on **ADD DESTINATION** in the upper right corner.

2. Choose either **Connect your destination** or **I don't have one**. For this tutorial you can either connect to an existing data warehouse you have access to by following the [relevant Fivetran documentation](https://fivetran.com/docs/databases) or use a [Fivetran-managed BigQuery service](https://fivetran.com/docs/destinations/bigquery/managed-bigquery). In this tutorial we will use Fivetrans managed BigQuery service by selecting **I don't have one** and then clicking on **CONTINUE SETUP**.

    ![Fivetran Destination](/img/guides/fivetran_destination.png)

:::info

> If you choose to connect to an existing data warehouse make sure you are able to connect a BI tool to it if you want to complete the optional Step 8. In Step 8 we will use [Tableau](https://www.tableau.com/) to visualize our data but you are free to use any BI tool compatible with your data warehouse.

Reviewer 2: I think this note might be better placed within Step 8 (and maybe reworded a bit accordingly)

:::

> 3. Configure your destination connection. You can choose the configuration freely. Astronomer recommends to use UTC as your timezone in all data tools as a best practice to. Click **SAVE & TEST** to save your destination.

Reviewer 2: ```suggestion
3. Configure your destination connection. You can choose any configuration. Astronomer recommends using UTC as your timezone in all data tools as a best practice. Click **SAVE & TEST** to save your destination.
```

    ![Fivetran-managed BigQuery](/img/guides/fivetran_managed_bigquery.png)

4. Click **Continue** to get back to the list of your destinations.

    ![Fivetran-managed BigQuery connection successful](/img/guides/fivetran_managed_bigquery_conn_success.png)

> ## Step 4: Connect Fivetran to a connector

Reviewer 2: Could this step be titled "Configure a Fivetran connector"? "Connect to a connector" sounds really weird to me, but I understand if that's the proper terminology :)

Fivetran needs at least one connector to be configured in order to create sync jobs. A connector is one of many data sources. View the Fivetran website for an [up-to-date list of connectors](https://www.fivetran.com/connectors).

1. Click on **Connectors** in the sidebar on the left side of the screen and then on **ADD CONNECTOR** in the upper right corner.

2. Select the GitHub Connector and click on **CONTINUE SETUP**.

3. Configure the GitHub Connector. Name the schema `in_github`. To connect Fivetran to GitHub you can use either [OAuth or a Personal Access Token](https://fivetran.com/docs/applications/github/setup-guide). Choose your method of authentication and click **AUTHORIZE**. In this tutorial we use OAuth authentication. 

    ![GH Connector](/img/guides/github_connector.png)

4. Authenticate Fivetran to your Github Account.

    ![Authorize Fivetran to GH](/img/guides/authorize_fivetran_github.png)

> 5. Select the repository `your-github-username/airflow-fivetran-tutorial` repository to be synced. You don't need to enable Webhooks for this tutorial.

Reviewer 2: ```suggestion
5. Select the `your-github-username/airflow-fivetran-tutorial` repository to be synced. You don't need to enable Webhooks for this tutorial.
```

    ![Select tutorial repository](/img/guides/fivetran_select_tutorial_repo.png)

6. Click **SAVE & TEST**. After your connection has been tested click **CONTINUE**.

> 7. Click on **Start Sync** to start your initial sync. This will extract metadata about your GitHub repository and load it into your destination. Once the initial sync is completed your Fivetran job is active, you can set the sync frequencies under the **Setup** tab.

Reviewer 2: ```suggestion
7. Click on **Start Sync** to start your initial sync. This will extract metadata about your GitHub repository and load it into your destination. Once the initial sync is completed and your Fivetran job is active, you can set the sync frequencies under the **Setup** tab.
```
I wonder if it makes sense to note somewhere why the initial sync has to be completed in Fivetran and not from the Airflow DAG?

    ![Initial sync completed](/img/guides/complete_initial_sync.png)

## Step 5: Generate a Fivetran API key

> We have now created a Fivetran sync job that will extract new metadata from a GitHub repository and load it into a relational database on a time-based schedule. But what if we want to run the sync job every time a specific event happens in our data ecosystem? This is where orchestration via Airflow can help Fivetran sync jobs to become event-driven. 

Reviewer 2: ```suggestion
You have now created a Fivetran sync job that will extract new metadata from a GitHub repository and load it into a relational database on a time-based schedule. But to run the sync job every time a specific event happens in your data ecosystem, you will need to use Airflow for orchestration. 
```

> In order to connect Airflow to Fivetran you will need a Fivetran API key. 

Reviewer 2: ```suggestion
In order to connect Airflow to Fivetran, you will need a Fivetran API key. 
```

> 1. Click on you account name in the right Fivetran sidebar and then on **API Key**. 

Reviewer 2: ```suggestion
1. Click on your account name in Fivetran's sidebar and then on **API Key**. 
```

    ![Fivetran generate API key](/img/guides/fivetran_generate_api_key.png)

2. Click **Generate API Key** and make sure to store the information displayed in a secure place.

## Step 6: Create an Airflow connection to Fivetran

1. Navigate to your Airflow UI at `localhost:8080`.

2. Click on **Admin** -> **Connections** -> **+** to create a new connection.

3. Name your connection `fivetran_conn` and select the **Fivetran** connection type. Provide your Fivetran API Key and Fivetran API Secret. If the connection type does not show up, double check that you added the Fivetran provider to your `requirements.txt` file prior to last starting your Airflow instance.

    ![Fivetran generate API key](/img/guides/fivetran_airflow_connection.png)

4. Click **Test** to test your connection. Once you see `Connection tested successfully` click **Save**.

## Step 5: Create your Airflow DAG

1. Open your `astro-fivetran-project` in a code-editor.

2. In your `dags` folder add a new Python file called `my_fivetran_dag.py`. 

3. Copy and past the following DAG code into the file:

>     <CodeBlock language="python">{fivetran_tutorial_dag_1}</CodeBlock>

Reviewer 2: After reading this through, I do think we should update to use the github sensor you mentioned. It's fine that there isn't an async version, because we can use the async version of the Fivetran operator in the main example DAG and not have a separate section for that. (Also maybe eventually there will be an async version and then we can just update the guide 😄 )

This DAG contains two tasks: 

> - the `wait_for_file` task uses the [FileSensor](https://registry.astronomer.io/providers/apache-airflow/modules/filesensor) to wait for a file to drop at a specific file path in the environment configured in the connection provided to `fs_conn_id`. In Step 1 you already added a connection with the ID `in_docker_file_conn` that connects to the Airflow environment within the Docker file. The `filepath` parameter in the copied DAG code uses a [Jinja template](templating.md) to wait for a csv file with a name starting with the logical date of the DAG in the format `yyyy-mm-dd` to be created in the `include` folder.

Reviewer 2: Let's update both of these bullets when the DAG is updated, and include some context about async operators in the second bullet.

4. Click **Test** to test your connection. Once you see `Connection tested successfully` click **Save**.

## Step 7: Create an Airflow connection to GitHub

The DAG you are about to build will wait for a specific tag to be added to a GitHub repository. In order for this to be possible you will need to add an Airflow connection to GitHub.

1. Click on **Admin** -> **Connections** -> **+** to create a new connection.

2. Name your connection `github_conn` and select the **GitHub** connection type. Provide a [Personal Access Token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) to your GitHub account. The PAT only needs to have one permission: `Read-only` access to the Metadata, which is the lowest level of access available to a PAT. You can restrict access to the `airflow-fivetran-tutorial` repository by using the `fine-grained personal access token` feature.

    ![GitHub connection](/img/guides/github_connection.png)

4. Click **Test** to test your connection. Once you see `Connection tested successfully` click **Save**.

## Step 8: Create your Airflow DAG

> 

Reviewer 2: Let's add a sentence up front here that describes overall what the DAG does. Something like "For this tutorial you will create a DAG that waits for a GitHub tag to be present in your GitHub repo, and then kicks off a Fivetran sync job to ingest the GitHub repo metadata."

1. Open your `astro-fivetran-project` in a code-editor.

2. In your `dags` folder add a new Python file called `my_fivetran_dag.py`. 

3. Copy and past the following DAG code into the file:

    <CodeBlock language="python">{fivetran_tutorial_dag_1}</CodeBlock>

This DAG contains four tasks: 

> - the `generate_tag_to_await` task uses the `@task` decorator to turn a Python function into an Airflow task. This task will pull a number from the [Airflow Variable](https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html) `sync-metadata` or create the Airflow Variable and assign it to `1` if it does not exist yet. The task returns the tag to wait for in the format `sync-metadata/{number}`. 

Reviewer 2: ```suggestion
- The `generate_tag_to_await` task pulls a number from the [Airflow Variable](https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html) `sync-metadata`, or create sthe Airflow Variable and assigns a value of `1` if it does not exist yet. The task returns the tag to wait for in the format `sync-metadata/{number}`. This task helps set up the GitHubTagSensor and ensures that DAG can be run repeatedly.
```
Feel free to adjust the last sentence, just wanted to add some context for why the first/last tasks are here.

> - the `wait_for_tag` task uses the [GitHubTagSensor](https://registry.astronomer.io/providers/github/modules/githubtagsensor) to wait for a specific tag to be present in a GitHub repository. In the example DAG we wait for `sync-metadata/{number}` to appear in the same repository that has been connected to the Fivetran job but you could also wait for a tag to be added in a different repository that the PAT you used in [Step 7](#step-7-create-an-airflow-connection-to-github) has access to.

Reviewer 2: ```suggestion
- The `wait_for_tag` task uses the [GitHubTagSensor](https://registry.astronomer.io/providers/github/modules/githubtagsensor) to wait for a specific tag to be present in your GitHub repository. This DAG is set up to wait for `sync-metadata/{number}` to appear in the repository that has been connected to the Fivetran job. 
```
Let's add a sentence at the end that says something like "If you want to wait for a tag in a different repository, update the XYZ parameter of the sensor", or whatever is relevant.

> - the `kick_off_fivetran_sync` task uses the FivetranOperatorAsync to kick of the Fivetran connector specified as `FIVETRAN_CONNECTOR_ID` as soon as the `wait_for_tag` task has completed successfully. This task will wait in a deferred state, not taking up an Airflow worker, until the Fivetran sync job has been completed successfully.

Reviewer 2: ```suggestion
- The `kick_off_fivetran_sync` task uses the FivetranOperatorAsync to kick of the Fivetran connector specified as `FIVETRAN_CONNECTOR_ID` as soon as the `wait_for_tag` task has completed successfully.
```

> - the `increase_tag_var` task will only run after `kick_off_fivetran_sync` has completed successfully and increases the value assigned to the `sync-metadata` Airflow Variable by 1.

Reviewer 2: ```suggestion
- The `increase_tag_var` task increases the value assigned to the `sync-metadata` Airflow Variable by 1. This task ensures that next time you run the DAG, the sensor will wait for a new tag value to be added to your GitHub repo.
```
Again, feel free to adjust the language I added here as necessary. Just wanted to add some context since this task isn't really part of the presented "use case"

> 4. Add the `FIVETRAN_CONNECTOR_ID` of your connector. You can find the ID of your connector in the Fivetran UI under the **Setup** tab:

Reviewer 2: ```suggestion
4. Update the `FIVETRAN_CONNECTOR_ID` variable with the ID of your connector. You can find the ID of your connector in the Fivetran UI under the **Setup** tab:
```
Fivetran connector IDs aren't considered confidential, are they?

Author: Haha, I had a moment asking myself the same thing when making the screenshot with mine. No, they are like deployment names or dag/task ids. :)

    ![Tableau BQ connection](/img/guides/tableau_bq_connection.png)

5. You can now visualize all elements of the GitHub repository metadata that has been loaded into BigQuery. For example, using the `commit` table, drag and drop the `commit(Count)` item from the Tables section to the `Rows` list and the `Committer Date` to the `Columns` list. Change the grain of `Committer Date` displayed to `Minute` by Clicking on the right side of the element in the `Columns` list. This will create a line graph showing how many commits were made every minute in the `airflow-fivetran-tutorial` repository. The screenshot below shows that two commits were made in separate minutes:

    ![Tableau commits 1](/img/guides/tableau_commits_one.png)

6. Add a few commits to your `airflow-fivetran-tutorial` repository.

7. Rerun the Fivetran sync job by rerunning the Airflow DAG and adding the GitHub tag the `wait_for_tag` task is listening to to your `airflow-fivetran-tutorial` repository.

8. View your updated dashboard with commits per minute of the hour aggregated.

    ![Tableau commits 1](/img/guides/tableau_commits_two.png)

## Conclusion

> Using Airflow with Fivetran allows you to embed your Fivetran jobs in your larger data ecosystem, making the scheduling of Fivetran jobs event- and data-driven. The new [`airflow-provider-fivetran-async`](https://registry.astronomer.io/providers/fivetran) offers new async capabilities to make your architecture more efficient.

Reviewer 2: ```suggestion
Using Airflow with Fivetran allows you to embed your Fivetran jobs in your larger data ecosystem, making the scheduling of Fivetran jobs event- and data-driven. The [`airflow-provider-fivetran-async`](https://registry.astronomer.io/providers/fivetran) offers asynchronous capabilities to make your architecture more scalable and efficient.
```
Just removing "new" since it won't be "new" for very long :)

1. Click on **Admin** -> **Connections** -> **+** to create a new connection.

2. Name your connection `github_conn` and select the **GitHub** connection type. Provide a [Personal Access Token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) (PAT) for your GitHub account. The PAT needs to have read access to the metadata (includes tags) of the `airflow-fivetran-tutorial` repository.

    ![GitHub connection](/img/guides/github_connection.png)

4. Click **Test** to test your connection. Once you see `Connection tested successfully` click **Save**.

## Step 8: Create your Airflow DAG

For this tutorial you will create a DAG that waits for a GitHub tag to be present in your GitHub repo, and then kicks off the Fivetran sync you configured to ingest the GitHub repo metadata to your destination.

1. Open your `astro-fivetran-project` in a code-editor.

2. In your `dags` folder add a new Python file called `my_fivetran_dag.py`. 

3. Copy and past the following DAG code into the file:

    <CodeBlock language="python">{fivetran_tutorial_dag_1}</CodeBlock>

> This DAG contains four tasks: 

Reviewer: This line and the following bullet points need to be indented by 4 spaces

- The `kick_off_fivetran_sync` task uses the FivetranOperatorAsync to kick of the Fivetran connector specified as `FIVETRAN_CONNECTOR_ID` as soon as the `wait_for_tag` task has completed successfully.
- The `increase_tag_var` task increases the value assigned to the `sync-metadata` Airflow Variable by 1. This task ensures that next time you run the DAG, the sensor will wait for a new tag value to be added to your GitHub repo.

4. Update the `FIVETRAN_CONNECTOR_ID` variable with the ID of your connector. You can find the ID of your connector in the Fivetran UI under the **Setup** tab:

    ![Fivetran connector ID](/img/guides/fivetran_connector_id.png)

5. Add your GitHub username to the `GITHUB_REPOSITORY` variable.

6. Save your DAG file with the changed variable names.

:::info

The FivetranOperatorAsync is one of many [deferrable operators](deferrable-operators.md). Instead of taking up a worker slot, these operators will hand their task to the Airflow Triggerer component while waiting for a condition to be fulfilled. For longer running tasks, this can result in cost savings and greater scalability as more worker slots area available.

:::

## Step 9: Run your DAG

> 1. In the Airflow UI, unpause your DAG to start its last scheduled DAG run with a logical date of yesterday.

Reviewer: Do you also have to toggle it on?

Reviewer: Sorry should have been more clear here, I mean because `catchup = False` do they also have to trigger a DAG run manually to get the sync to work, now that it's on a daily schedule?

Author: Ah, not for the first run if the start_date is in the past. Airflow will always backfill the last run that would have happend the moment you unpause the DAG for the first time, even if catchup is False. I find it a bit confusing personally, I often forget, hit the manual run and then end up with 2 DAG runs at the same time 😅

2. Once the `wait_for_tag` task is running check the task logs to see what `tag` in which repository the task is waiting for. It should be `sync-metadata/1` in `<your GH name>/airflow-fivetran-tutorial`.

    ![GH tag sensor waiting](/img/guides/fivetran_githubtagsensor.png)

3. Add a tag with the name of `sync-metadata/1` to `<your GH name>/airflow-fivetran-tutorial`. You can either [use the GitHub UI to add tags](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/managing-tags) or run the following bash commands in your local repository clone:

    ```sh
    git tag sync-metadata/1
    git push --tags
    ```

> 4. After a few seconds the sensor will detect the added file and print logs similar to the ones shown below.

Reviewer: This isn't a user action, remove the number and have it as supplemental info to step 3? Or reframe it as "Check your logs again"

    ```text
    [2023-03-01, 00:31:35 UTC] {custom_githubtagsensor.py:127} INFO - Poking for tag: sync-metadata/1 in repository: TJaniF/airflow-fivetran-tutorial
    [2023-03-01, 00:31:35 UTC] {base.py:73} INFO - Using connection ID 'github_conn' for task execution.
    [2023-03-01, 00:31:36 UTC] {custom_githubtagsensor.py:146} INFO - Tag sync-metadata/1 exists in TJaniF/airflow-fivetran-tutorial repository, Success.
    [2023-03-01, 00:31:36 UTC] {base.py:228} INFO - Success criteria met. Exiting.
    [2023-03-01, 00:31:36 UTC] {taskinstance.py:1318} INFO - Marking task as SUCCESS. 
    ```

5. After the `wait_for_tag` task has completed, the `kick_off_fivetran_sync` task will start. This task will be in a deferred state (violet square) until the Fivetran sync is completed. 

    ![Fivetran Deferred](/img/guides/fivetran_deferred_operator.png)

> 

Reviewer: Remove this step if there's no user action

6. After the DAG run has finished successfully, verify in your Fivetran UI that the sync shows an additional `Manual Update Triggered` entry in its `User Actions`.

    ![Fivetran additional sync](/img/guides/fivetran_additional_sync.png)

:::info

> When running the DAG again after it has run successfully the `number` Variable will have increased by one, meaning in its second run, the `wait_for_tag` task will wait for a tag with the name `sync-metadata/2` and so forth. To reset the Airflow Variable manually go to the Airflow UI and navigate to **Admin** -> **Variables**.

Reviewer: ```suggestion
When you run this DAG again, the `TAG_NAME` variable will have incremented by one, meaning that the `wait_for_tag` task will wait for a tag with the name `sync-metadata/2` and so forth. To reset `TAG_NAME` manually, open the Airflow UI and update the variable value in **Admin** -> **Variables**.
```

:::

## (Optional) Step 10: Visualize your commits with Tableau

> So far this tutorial has shown you how to integrate Fivetran and Airflow but we haven't had a look at the data we've been running through the pipeline. If you are using a custom Fivetran Destination you can simply view your data directly and connect your favorite BI tool to it. If you followed this tutorial using Fivetran's managed BigQuery service, you can add a BI tool in order to view your data!

Reviewer: ```suggestion
As a bonus, you can implement data observability into your data pipeline. If you are using a custom Fivetran destination, you can simply view your data directly and connect your favorite BI tool to it. If you followed this tutorial using Fivetran's managed BigQuery service, you can add a BI tool like Tableau to view your data.
```

> :::info
> 
> If you choose to connect to an existing data warehouse not managed by Fivetran, you can directly connect a BI tool to it to complete this step. We will use [Tableau](https://www.tableau.com/) to visualize our data but you are free to use any BI tool compatible with your data warehouse.
> 
> :::

Reviewer: ```suggestion
```
This is duplicative of the content in the intro

> 1. In the Fivetran UI navigate to **Destinations**, click on your warehouse and then select the **BI tools** tab and click on **+ BI tool**.

Reviewer: ```suggestion
1. In the Fivetran UI go to **Destinations** and select your warehouse, click the **BI tools** tab, and click **+ BI tool**.
```

> 2. You can use [any compatible BI tool](https://fivetran.com/docs/destinations/bigquery/managed-bigquery#addbitools) listed that you are familiar with. This tutorial will use Tableau. Follow the [steps in Fivetrans documentation to authenticate Tableau](https://fivetran.com/docs/destinations/bigquery/managed-bigquery/tableau-setup-guide#spanclassstepitemauthorizegoogleuserintableauspan) to the Fivetran-managed BigQuery service. Note that you have to authorize the same Google user to Tableau as you authorized to the Fivetran-managed BigQuery service.

Reviewer: ```suggestion
2. Authenticate to Tableau. See [Fivetran documentation](https://fivetran.com/docs/destinations/bigquery/managed-bigquery/tableau-setup-guide#spanclassstepitemauthorizegoogleuserintableauspan). Note that you have to authorize the same Google user to Tableau as you authorized to the Fivetran managed BigQuery service.
```

>     ![Fivetran Tableau connection](/img/guides/fivetran_tableau_connection.png)
> 

Reviewer: ```suggestion
```

> 3. [Create a new Tableau worksheet](https://help.tableau.com/current/pro/desktop/en-us/environment_workspace.htm), either in Tableau desktop or in Tableau Cloud. This tutorial shows screenshots using Tableau Cloud.

Reviewer: ```suggestion
3. [Create a new Tableau worksheet](https://help.tableau.com/current/pro/desktop/en-us/environment_workspace.htm) either in Tableau desktop or in Tableau Cloud.
```

> 4. Connect your Tableau worksheet to the BigQuery project created by the Fivetran-managed service. Make sure you are using the same Google user you authorized in Fivetran. See the [Tableau documentation](https://help.tableau.com/current/pro/desktop/en-us/examples_googlebigquery.htm) for help.

Reviewer: ```suggestion
4. Connect your Tableau worksheet to the BigQuery project created by the Fivetran managed service. Make sure you are using the same Google user you authorized in Fivetran. See [Tableau documentation](https://help.tableau.com/current/pro/desktop/en-us/examples_googlebigquery.htm).
```

> 
>     ![Tableau BQ connection](/img/guides/tableau_bq_connection.png)

Reviewer: ```suggestion
```

> 5. You can now visualize all elements of the GitHub repository metadata that has been loaded into BigQuery. For example, using the `commit` table, drag and drop the `commit(Count)` item from the Tables section to the `Rows` list and the `Committer Date` to the `Columns` list. Change the grain of `Committer Date` displayed to `Minute` by Clicking on the right side of the element in the `Columns` list. This will create a line graph showing how many commits were made every minute in the `airflow-fivetran-tutorial` repository. The screenshot below shows that two commits were made in separate minutes:

Reviewer: ```suggestion
5. You can now visualize all elements of the GitHub repository metadata that has been loaded into BigQuery. Using the **commit** table, drag and drop the **commit(Count)** item from the Tables section to the `Rows` list and the **Committer Date** to the `Columns` list. Change the grain of **Committer Date** displayed to **Minute** by clicking on the right side of the element in the **Columns** list. This creates a line graph showing how many commits were made each minute in the `airflow-fivetran-tutorial` repository. The following screenshot shows that two commits were made in separate minutes:
```

Reviewer: Screen elements are always bolded, we only use code styling for user inputs or code

    ![Tableau commits 1](/img/guides/tableau_commits_one.png)

6. Add a few commits to your `airflow-fivetran-tutorial` repository.

> 7. Rerun the Fivetran sync by rerunning the Airflow DAG and adding the GitHub tag the `wait_for_tag` task is listening to to your `airflow-fivetran-tutorial` repository.

Reviewer: ```suggestion
7. Rerun the Fivetran sync as described in [Step 9](#step-9-run-your-airflow-dag).
```
